server/src-lib/Hasura/RQL/DDL/Schema.hs

```diff
@@ -58,22 +61,26 @@ data RunSQL
   { rSql                      :: Text
   , rCascade                  :: !(Maybe Bool)
   , rCheckMetadataConsistency :: !(Maybe Bool)
+  , rReadOnly                 :: !(Maybe Bool)
   } deriving (Show, Eq, Lift)
 $(deriveJSON (aesonDrop 1 snakeCase){omitNothingFields=True} ''RunSQL)
 
 runRunSQL :: (CacheBuildM m, UserInfoM m) => RunSQL -> m EncJSON
-runRunSQL (RunSQL t cascade mChkMDCnstcy) = do
+runRunSQL (RunSQL sql cascade mChkMDCnstcy (fromMaybe False -> isReadOnly)) = do
   adminOnly
-  isMDChkNeeded <- maybe (isAltrDropReplace t) return mChkMDCnstcy
-  bool (execRawSQL t) (withMetadataCheck (or cascade) $ execRawSQL t) isMDChkNeeded
+  isMDChkNeeded <- maybe (if isReadOnly then pure False else isAltrDropReplace sql) return mChkMDCnstcy
+  bool (execRawSQL sql) (withMetadataCheck (or cascade) $ execRawSQL sql) isMDChkNeeded
```

Comment:
Maybe it’s just me, but I feel like all the different ways of handling boolean logic here are getting a little confusing. Personally, I think a lot of this code would read a lot more clearly if we wrote a manual `FromJSON RunSQL` instance that handles the default values there instead of having `Maybe`s. Then maybe we could avoid so much of the twisty logic in this function and just do something like this:

```haskell
isMetadataCheckNeeded <- maybeCheckMetadataConsistency `onNothing` if
  | isReadOnly -> pure False
  | otherwise  -> isAlterDropReplace sql
```

But then I start wondering: is the current logic even right? Does it make sense to not check metadata consistency if the user specifies `check_metadata_consistency: false` in the query, but they also specify `cascade: true`? That seems nonsensical to me, but maybe I’m missing something.

Reply:
what do you think if we use `ViewPatterns` to peel the maybe with default? (Just like mReadOnly). Writing `(fromMaybe False -> isReadOnly)` is not very clean but we can wrap that like `(setDefaultFalse -> isReadOnly)` for more cleanliness.

Reply:
I’d prefer doing it in the `FromJSON` instance so that we normalize things at the same time we parse them; it’s really a parsing concern and it avoids the kind of mistake where the logic is duplicated in two different places in different ways (admittedly pretty unlikely here, but a good principle to follow in general, imo).